### PR TITLE
Kmd client

### PR DESCRIPTION
**Pull request by amityadav0**

### CODE REVIEW

File: examples/deployment/algob.config.js [hunk 1]
```diff
@@ -52,6 +53,28 @@ let accounts = mkAccounts([
   }
 ]);
 
+// KMD credentials
+let KMDConfig = {
```
Comment (Invertisment) [inline examples/deployment/algob.config.js:57]: You call this KMDConfig but add it later as a network.
Isn't it that the network uses one port and KMD has it's own?

File: packages/algob/sample-project/algob.config.js [hunk 2]
```diff
@@ -31,6 +54,7 @@ let defaultCfg = {
 module.exports = {
   networks: {
     localhost: defaultCfg,
-    default: defaultCfg
+    default: defaultCfg,
+    KMD: KMDConfig
```
Comment (Invertisment) [inline packages/algob/sample-project/algob.config.js:58]: You add KMDConfig as a network but it shouldn't be used as a node that can compile code.

File: packages/algob/src/types.ts [hunk 3]
```diff
@@ -34,6 +34,11 @@ export type AccountDef =
   | HDAccount
   | Account;
 
+export interface wallet {
```
Comment (Invertisment) [inline packages/algob/src/types.ts:37]: Change `wallet` to `Wallet`, it's a type

File: examples/deployment/algob.config.js (hunk 1 above)
Comment (amityadav0) [reply examples/deployment/algob.config.js:57]: yes, i was trying to include it in env, so that KMD token is accessible from env.

Comment (robert-zaremba) [reply examples/deployment/algob.config.js:57]: One more thing for same line: local variables should be lower case (unless we define exported constants - which then should be upper case)

File: packages/algob/src/types.ts (hunk 3 above)
Comment (amityadav0) [reply packages/algob/src/types.ts:37]: ok

File: examples/deployment/algob.config.js [hunk 4]
```diff
@@ -52,6 +53,28 @@ let accounts = mkAccounts([
...
+
+// load KMD Accounts from given wallet - (addr, secret key)
+async function kmdAdd() {
+  const kmdAddresses = await loadKMDAddresses(KMDConfig.host, KMDConfig.token, KMDConfig.port, 
+    wallet.name, wallet.password);
+
+    console.log(kmdAddresses);
+}
+
+kmdAdd();
```
Comment (robert-zaremba) [inline examples/deployment/algob.config.js:76]: here is the problem. You don't wait for the `kmdAdd` competition. This function is asynchronous.

What you can do is:

1. Extend the network type definition and add `kmdAddresses: Promise<AccountSDK[]>` (provisionarily)
1. assign kmd addresses to `default` network:

        defaltCfg.kmdAddresses = loadKMDAddresses(KMDConfig.host, KMDConfig.token, KMDConfig.port, 
    wallet.name, wallet.password);

1. Print kmdAddressess in the `node-info` comand (builtin-tasks).  -- to print, you will need to await first:

        let kmdaddrs = await network.kmdAddresses;

File: packages/algob/sample-project/algob.config.js (hunk 2 above)
Comment (robert-zaremba) [reply packages/algob/sample-project/algob.config.js:58]: correct.

File: packages/types-algosdk/index.d.ts [hunk 5]
```diff
@@ -20,6 +20,42 @@ declare module 'algosdk' {
     statusAfterBlock(lastround: number): Action<any>
   }
 
+ class Kmd {
```
Comment (robert-zaremba) [inline packages/types-algosdk/index.d.ts:23]: please call this class `KMD`

File: packages/types-algosdk/index.d.ts [hunk 6]
```diff
@@ -20,6 +20,42 @@ declare module 'algosdk' {
...
+    
+    versions(): Promise<any>;
+    listWallets(): Promise<wallet>;
+    initWalletHandle(walletid: string, password: string): Promise<walletHandle>
+    listKeys(wallet_handle_token: string): Promise<Keys>;
+    exportKey(wallet_handle_token: string, Password: string, address: string): 
+    Promise<privateKey>
+  }
+
+  interface wallet {
```
Comment (robert-zaremba) [inline packages/types-algosdk/index.d.ts:34]: the interface name should be `wallets`

File: packages/types-algosdk/index.d.ts [hunk 7]
```diff
@@ -20,6 +20,42 @@ declare module 'algosdk' {
...
+    private_key: Uint8Array
+  }
+
+  interface Keys {
+    addresses: string[]
+  }
+
+  interface walletHandle {
+    wallet_handle_token: string
+  }
```
Comment (robert-zaremba) [inline packages/types-algosdk/index.d.ts:57]: All interface names should be Upper Case

File: packages/types-algosdk/index.d.ts [hunk 8]
```diff
@@ -91,7 +127,7 @@ declare module 'algosdk' {
 
   export function Indexer (...args: any[]): any;
 
-  export function Kmd(token: any, baseServer: any, port: any): any;
+  //export function Kmd(token: any, baseServer: any, port: any): any;
```
Comment (robert-zaremba) [inline packages/types-algosdk/index.d.ts:130]: let's remove this line.

File: examples/deployment/algob.config.js (hunk 4 above)
Comment (robert-zaremba) [reply examples/deployment/algob.config.js:76]: I mean, you would need to call `await kmdAdd()` to have this working.

File: packages/types-algosdk/index.d.ts (hunk 5 above)
Comment (amityadav0) [reply packages/types-algosdk/index.d.ts:23]: ok

File: packages/types-algosdk/index.d.ts (hunk 6 above)
Comment (amityadav0) [reply packages/types-algosdk/index.d.ts:34]: OK

File: packages/types-algosdk/index.d.ts (hunk 5 above)
Comment (amityadav0) [reply packages/types-algosdk/index.d.ts:23]: cannot change that as it is in algosdk module

File: packages/types-algosdk/index.d.ts (hunk 8 above)
Comment (amityadav0) [reply packages/types-algosdk/index.d.ts:130]: OK

File: examples/deployment/algob.config.js [hunk 9]
```diff
@@ -67,6 +90,7 @@ let defaultCfg = {
 module.exports = {
   networks: {
     localhost: defaultCfg,
-    default: defaultCfg
+    default: defaultCfg,
+    KMD: KMDConfig
```
Comment (robert-zaremba) [inline examples/deployment/algob.config.js:94]: this should be a part of a network configuration.

File: packages/algob/sample-project/algob.config.js [hunk 10]
```diff
@@ -8,29 +8,43 @@
...
 // const accFromFile = loadAccountsFromFileSync("assets/accounts_generated.yaml");
 // accounts = accounts.concat(accFromFile);
-
+//console.log("OK");
 let defaultCfg = {
   host: "http://localhost",
   port: 8080,
   token: "content_of/algorand-node-data/algod.token",
   accounts: accounts,
+  kmdAddresses: await loadKMDAddresses(kmdconfig.host, kmdconfig.token, kmdconfig.port,
```
Comment (robert-zaremba) [inline packages/algob/sample-project/algob.config.js:41]: do we need to use `await` here?

If `kmdAddresses` will be a promise then it's ok, and we will not have any `async / await` in the config and everything will be simpler.

File: packages/algob/src/lib/account.ts [hunk 11]
```diff
@@ -71,3 +71,23 @@ export function mkAccountIndex (accountList: Account[]): Accounts {
   }
   return out;
 }
+
+export async function loadKMDAddresses (host: string, token: string, port: number,
```
Comment (robert-zaremba) [inline packages/algob/src/lib/account.ts:75]: please rename this function to `loadKMDAccounts`.

File: packages/algob/src/lib/account.ts [hunk 12]
```diff
@@ -71,3 +71,23 @@ export function mkAccountIndex (accountList: Account[]): Accounts {
...
+  let walletid = '';
+  const wallets = (await kmdclient.listWallets()).wallets;
+  for (const arrayItem of wallets) {
+    if (arrayItem.name === walletName) {
+      walletid = arrayItem.id;
+    }
+  }
+  const wallethandle = (await kmdclient.initWalletHandle(walletid, password)).wallet_handle_token;
+  const address = await kmdclient.listKeys(wallethandle);
+  const mkaccount: AccountSDK[] = [];
```
Comment (robert-zaremba) [inline packages/algob/src/lib/account.ts:87]: why this variable is called `mkaccount`?

File: packages/algob/test/helpers/environment.ts [hunk 13]
```diff
@@ -64,7 +64,7 @@ export function getEnv (defaultNetworkCfg?: NetworkConfig): AlgobRuntimeEnv {
     debug.enable("algob*");
   }
 
-  const config = loadConfigAndTasks(runtimeArgs);
+  const config = await loadConfigAndTasks(runtimeArgs);
```
Comment (robert-zaremba) [inline packages/algob/test/helpers/environment.ts:67]: again, if we won't use `async / await` in config, then everything will be simpler.

File: packages/algob/test/internal/core/config/config-loading.ts [hunk 14]
```diff
@@ -68,32 +68,36 @@ describe("config loading", function () {
...
+
+        if (!config.paths) {
+          assert.fail("Project was not loaded");
+        }
+
+        assert.equal(
+          config.paths.configFile,
+          path.normalize(path.join(process.cwd(), "config.js"))
+        );
+      })().catch((err) => console.log(err));
```
Comment (robert-zaremba) [inline packages/algob/test/internal/core/config/config-loading.ts:100]: don't do this. You can make the function `async`:
```
 it("should accept an absolute path", async function () {
```
Mocha works well with async functions.

Comment (robert-zaremba) [reply packages/algob/test/internal/core/config/config-loading.ts:100]: same in other tests. 

But, if we will be ok, that in the condig, we will not use `await` then we don't need this changes.

File: packages/algob/test/internal/core/config/config-resolution.ts [hunk 15]
```diff
@@ -25,36 +25,44 @@ describe("Config resolution", () => {
...
-        assert.isTrue(algobChainCfg.throwOnTransactionFailures);
-        assert.isTrue(algobChainCfg.throwOnCallFailures);
+        (async () => {
+          const config = await loadConfigAndTasks();
+          assert.containsAllKeys(config.networks, [ALGOB_CHAIN_NAME]);
+
+          const algobChainCfg = config.networks[ALGOB_CHAIN_NAME] as AlgobChainCfg;
+          assert.isTrue(algobChainCfg.throwOnTransactionFailures);
+          assert.isTrue(algobChainCfg.throwOnCallFailures);
+        })().catch((err) => console.log(err));
```
Comment (robert-zaremba) [inline packages/algob/test/internal/core/config/config-resolution.ts:35]: as above.

File: packages/algob/sample-project/algob.config.js (hunk 10 above)
Comment (robert-zaremba) [reply packages/algob/sample-project/algob.config.js:41]: Also, let's rename this to `kmdAccounts`

Comment (robert-zaremba) [reply packages/algob/sample-project/algob.config.js:41]: if you are using `await` then the value here is not promise (await resolve promises). If we don't use await then it should be fine, and we will do await in other parts of the code.

**Review by robert-zaremba — CHANGES_REQUESTED** [review]
let's try to remove async / await from the config.

File: packages/algob/sample-project/algob.config.js (hunk 10 above)
Comment (amityadav0) [reply packages/algob/sample-project/algob.config.js:41]: yes i did this without await first, but that didn't work, i was getting a blank array,
I added await config.kmdaddresess in node-info
and then ran node-info from cli

File: docs/algob-config.md [hunk 16]
```diff
@@ -47,6 +47,24 @@ Each network configuration requires a list of accounts. These accounts are then
           mnemonic: "call boy rubber fashion arch day capable one sweet skate outside purse six early learn tuition eagle love breeze pizza loud today popular able divide"
         }]);
 
+1. Loaded from Key Management Daemon (KMD)
+
+        // KMD credentials
+        let KMDConfig = {
```
[the remark this answers is not in the document]
Comment (amityadav0) [reply docs/algob-config.md:53]: ok

File: examples/deployment/algob.config.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (amityadav0) [reply examples/deployment/algob.config.js:57]: ok

File: examples/deployment/algob.config.js [hunk 17]
```diff
@@ -52,6 +53,28 @@ let accounts = mkAccounts([
...
 ]);
 
+// KMD credentials
+let KMDConfig = {
+  host: "127.0.0.1",
+  port: 7833,
+  token: "09c2da31d3e3e96ed98ba22cc4d58a14184f1808f2b4f21e66c9d38f70ca7232"
+}
+
+let wallet = {
```
[the remark this answers is not in the document]
Comment (amityadav0) [reply examples/deployment/algob.config.js:63]: ok

File: packages/algob/sample-project/algob.config.js [hunk 18]
```diff
@@ -8,29 +8,43 @@
...
+let kmdconfig = {
+  host: "127.0.0.1",
+  port: 7833,
+  token: "09c2da31d3e3e96ed98ba22cc4d58a14184f1808f2b4f21e66c9d38f70ca7232"
+}
+
+let wallet = {
+  name: "MyTestWallet1",
+  password: "testpassword"
+}
```
[the remark this answers is not in the document]
Comment (amityadav0) [reply packages/algob/sample-project/algob.config.js:30]: ok

File: packages/algob/sample-project/algob.config.js [hunk 19]
```diff
@@ -8,29 +8,43 @@
...
 // const { loadAccountsFromFileSync } = require("algob");
 // // Example: accounts loaded from a file:
 // const accFromFile = loadAccountsFromFileSync("assets/accounts_generated.yaml");
 // accounts = accounts.concat(accFromFile);
-
+//console.log("OK");
```
[the remark this answers is not in the document]
Comment (amityadav0) [reply packages/algob/sample-project/algob.config.js:35]: ok

File: examples/deployment/algob.config.js (hunk 9 above)
Comment (amityadav0) [reply examples/deployment/algob.config.js:94]: ok

File: packages/algob/sample-project/algob.config.js (hunk 10 above)
Comment (amityadav0) [reply packages/algob/sample-project/algob.config.js:41]: when we load the algob.config.js without resolving promise it stores a blank array in kmdAddresses, 
and even after adding await in node-info , it results in blank array

File: packages/algob/src/lib/account.ts (hunk 12 above)
Comment (amityadav0) [reply packages/algob/src/lib/account.ts:87]: ok, i will rename it to accounts.

File: packages/algob/test/internal/core/config/config-loading.ts (hunk 14 above)
Comment (amityadav0) [reply packages/algob/test/internal/core/config/config-loading.ts:100]: ok

File: packages/algob/test/internal/core/config/config-resolution.ts (hunk 15 above)
Comment (amityadav0) [reply packages/algob/test/internal/core/config/config-resolution.ts:35]: ok

**Pull request closed without merging**